core/src/main/scala/io/projectglow/sql/SqlExtensionProvider.scala

```diff
@@ -43,130 +49,88 @@ class GlowSQLExtensions extends (SparkSessionExtensions => Unit) {
+  }
 
-    functionRegistry.registerFunction(
-      FunctionIdentifier("hard_calls"),
-      exprs => {
-        if (exprs.size == 4) {
-          HardCalls(exprs(0), exprs(1), exprs(2), exprs(3))
+  /**
+   * Register SQL functions based on a yaml function definition file.
+   */
+  def registerFunctions(
+      conf: SQLConf,
+      functionRegistry: FunctionRegistry,
+      resourcePath: String = FUNCTION_YAML_PATH): Unit = {
+
+    loadFunctionDefinitions(resourcePath).foreach { _function =>
+      val function = _function.asScala
+      val id = FunctionIdentifier(function("name").asInstanceOf[String])
+      //      println(s"Building $id")
+      val exprClass = function("expr_class").asInstanceOf[String]
+      val args = function("args").asInstanceOf[JList[JMap[String, Any]]].asScala
+      val argsDoc = args.map { _arg =>
+        val arg = _arg.asScala
+        val suffix = if (arg.get("is_optional").exists(_.asInstanceOf[Boolean])) {
+          " (optional)"
+        } else if (arg.get("is_var_args").exists(_.asInstanceOf[Boolean])) {
+          " (repeated)"
         } else {
-          HardCalls(exprs(0), exprs(1), exprs(2))
+          ""
         }
-      }
-    )
-
-    functionRegistry.registerFunction(
-      FunctionIdentifier("hardy_weinberg"),
-      exprs => HardyWeinberg(exprs.head)
-    )
-
-    functionRegistry.registerFunction(
-      FunctionIdentifier("lift_over_coordinates"),
-      exprs => {
-        if (exprs.size == 5) {
-          LiftOverCoordinatesExpr(exprs(0), exprs(1), exprs(2), exprs(3), exprs(4))
-        } else {
-          LiftOverCoordinatesExpr(exprs(0), exprs(1), exprs(2), exprs(3))
+        s"${arg("name")}: ${arg("doc")} $suffix"
+      }.mkString("\n")
+      val info = new ExpressionInfo(
+        exprClass,
+        null,
+        function("name").asInstanceOf[String],
+        function("doc").asInstanceOf[String],
+        argsDoc,
+        "",
+        "",
+        function("since").asInstanceOf[String]
+      )
+      functionRegistry.registerFunction(
+        id,
+        info,
+        exprs => {
+          val clazz = Class.forName(exprClass, true, Thread.currentThread().getContextClassLoader)
+          val children = args.zipWithIndex.flatMap {
+            case (_arg: JMap[String, Any], idx: Int) =>
+              val arg = _arg.asScala
+              if (arg.get("is_optional").exists(_.asInstanceOf[Boolean]) && idx >= exprs.size) {
+                None
+              } else if (arg.get("is_var_args").exists(_.asInstanceOf[Boolean])) {
+                Some(exprs.slice(idx, exprs.size))
+              } else if (idx >= exprs.size) {
+                throw parameterError(id.funcName, exprs.size)
+              } else if (idx == args.size - 1 && exprs.size != args.size) {
+                throw parameterError(id.funcName, exprs.size)
+              } else {
+                Some(exprs(idx))
+              }
+          }
+          val constructor = clazz
+            .getConstructors
```

Comment:
To clarify, does `getConstructors` work with `this()`, but not with Scala constructors that provide default parameters?

Reply:
We're using Java reflection here, and it doesn't understand Scala features like default arguments.